### PR TITLE
Password Checking

### PR DESCRIPTION
**Pull request by abhisuri97**
This is a macro implementing zxcvbn password checking on the frontend. Takes in the id for the password field on which the developer wishes to do a strength check. 
Example Usage:
`{{ password_check('password-id-field-name-here`) }}`
Disables submit button until password strength of >= 2 is reached.

Example for the Change Password Page:
On `app/templates/account/manage.html` insert `{% import 'macros/check_password.html' as check %}` and place `{{ check.password_check('new_password') }}` in the line before `{% endblock %}`
Bad Password:
Bad Password

Strong Password:
Strong Password

### CODE REVIEW

**Comment by maxmcc** [comment]
From [CanIUse](http://caniuse.com/#search=meter) it doesn't seem like there's a great deal of support for the `<meter>` element in IE and Mobile Safari. At the very least we should do some testing to make sure there's graceful degradation (it would suck not to be able to see why the form isn't accepting your password).

**Comment by abhisuri97** [comment]
I'll look into <meter> and probably replace it with a div. However, i am not sure how to provide feedback since zxcvbn is kinda complicated. should i just put like password tips? (like "make sure its longer and has special characters?).

**Comment by sandlerben** [comment]
I like including this as a visual tool for users but I do **not** think we should block the use of weak passwords. Especially because zxcvbn judges strength on many different parameters which aren't transparent or easy to understand.

Maybe make requiring a certain strength an argument to the macro?

File: app/templates/macros/check_password.html [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+  meter[value="4"]::-webkit-meter-optimum-value { background: green; }
+
+  /* Gecko based browsers */
+  meter[value="1"]::-moz-meter-bar { background: red; }
+  meter[value="2"]::-moz-meter-bar { background: yellow; }
+  meter[value="3"]::-moz-meter-bar { background: orange; }
+  meter[value="4"]::-moz-meter-bar { background: green; }
+</style>
+
+{% endmacro %}
```
Comment (sandlerben) [inline app/templates/macros/check_password.html]: Space at end of file?

**Comment by sandlerben** [comment]
Besides that and what Max said, looks good.

**Comment by abhisuri97** [comment]
@maxmcc would the `<progress>` element be fine, it doesn't work on opera mini but everything else works fine as far as CanIUse says.

**Comment by maxmcc** [comment]
Yeah, assuming you can style it the way you want

**Comment by abhisuri97** [comment]
Changed to Progress Bar. Works in all major browsers (except IE 8 where it falls back to the unstyled progress bar but will still display progress). New syntax for calling `password_check('password-id-field-name-here', minimum-password-strength-0-thru-4)` e.g. `{{ password_check('password', 1) }}`

**Comment by sandlerben** [comment]
Can we change the password fields to use this macro as part of this PR?

**Comment by abhisuri97** [comment]
@sandlerben what default password strength do you want? (0 thru 4)

**Comment by sandlerben** [comment]
I think 0 (see my comment above).

**Comment by sandlerben** [comment]
@maxmcc thoughts?

**Comment by abhisuri97** [comment]
added the password check for all the forms I could find that required a new password. all defaults are at 0

**Comment by sandlerben** [comment]
Ok great. Let me just run the app and then I'll give a final :+1: Expect a response in the next few days (I'm very busy right now).

**Comment by abhisuri97** [comment]
next few days...

**Comment by sandlerben** [comment]
image

**Comment by sandlerben** [comment]
@abhisuri97 this looks good feel free to merge. In the future can you make changes on branches in the hack4impact/flask-base repo? It is easier for me to check out and run locally.

**Pull request merged**